src/main/kotlin/org/rust/cargo/toolchain/Rustup.kt

```diff
@@ -21,25 +23,40 @@ class Rustup(
     private val rustup: Path,
     private val projectDirectory: Path
 ) {
-    sealed class DownloadResult {
-        class Ok(val library: VirtualFile) : DownloadResult()
-        class Err(val error: String) : DownloadResult()
+    @Suppress("unused")
+    sealed class DownloadResult<out T> {
+        class Ok<T>(val value: T) : DownloadResult<T>()
+        class Err<T>(val error: String) : DownloadResult<T>()
```

Comment:
I released that we can define Err class as
`class Err(val error: String) : DownloadResult<Nothing>`

Reply:
Indeed:)
And probably, we want to implement a `Result` class for common purposes because we already have at least another two implementations: `TaskResult` and `ResolveResult`

Reply:
Maybe we want to use the library https://github.com/michaelbull/kotlin-result